sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/SchemaHelper.scala

```diff
@@ -0,0 +1,149 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.spark.sql.execution.streaming.state
+
+import java.io.StringReader
+
+import org.apache.hadoop.fs.{FSDataInputStream, FSDataOutputStream}
+
+import org.apache.spark.sql.execution.streaming.MetadataVersionUtil
+import org.apache.spark.sql.types.StructType
+import org.apache.spark.util.Utils
+
+/**
+ * Helper classes for reading/writing state schema.
+ */
+object SchemaHelper {
+
+  sealed trait SchemaReader {
+    def read(inputStream: FSDataInputStream): (StructType, StructType)
+  }
+
+  object SchemaReader {
+    def createSchemaReader(schemaVersion: String): SchemaReader = {
+      val version = MetadataVersionUtil.validateVersion(schemaVersion,
+        StateSchemaCompatibilityChecker.VERSION)
+      version match {
+        case 1 => new SchemaV1Reader
+        case 2 => new SchemaV2Reader
+      }
+    }
+  }
+
+  class SchemaV1Reader extends SchemaReader {
+    def read(inputStream: FSDataInputStream): (StructType, StructType) = {
+      val keySchemaStr = inputStream.readUTF()
+      val valueSchemaStr = inputStream.readUTF()
+      (StructType.fromString(keySchemaStr), StructType.fromString(valueSchemaStr))
+    }
+  }
+
+  class SchemaV2Reader extends SchemaReader {
+    def read(inputStream: FSDataInputStream): (StructType, StructType) = {
+      val buf = new StringBuilder
+      val numKeyChunks = inputStream.readInt()
+      (0 until numKeyChunks).foreach(_ => buf.append(inputStream.readUTF()))
+      val keySchemaStr = buf.toString()
+
+      buf.clear()
+      val numValueChunks = inputStream.readInt()
+      (0 until numValueChunks).foreach(_ => buf.append(inputStream.readUTF()))
+      val valueSchemaStr = buf.toString()
+      (StructType.fromString(keySchemaStr), StructType.fromString(valueSchemaStr))
+    }
+  }
+
+  trait SchemaWriter {
+    val version: String
```

Comment:
The representation of version won't be changed, so "v" can just be directly referred in `writeVersion` and this can be `Int`.